src/init/init.c

```diff
@@ -87,7 +87,27 @@ static inline void lp_sram_unpack(void)
 
 #if CONFIG_MULTICORE
 
-#ifndef __ZEPHYR__
+#ifdef __ZEPHYR__
+
+static int check_restore(void)
+{
+	struct idc *idc = *idc_get();
+	struct notify *notifier = *arch_notify_get();
+	struct schedulers *schedulers = *arch_schedulers_get();
+
+	/* check whether basic core structures has been already allocated. If they
+	 * are available in memory, it means that this is not cold boot and memory
+	 * has not been powered off.
+	 */
+	if (!idc || !notifier || !schedulers)
+		return 0;
```

Comment:
even though this function returns `int` (can we change it to return `bool`?) still `return idc && notifier && schedulers;` should still work. Also, seeing this check raises a question for me - what if some of these pointers are `NULL`? I think the answer is that this isn't possible, right? If any of these allocations fail initially, the whole initialisation fails, so we never reach this point. That means, that checking just one of these pointers should be enough: if one of them is set, all the others must be set too?

Reply:
> can we change it to return `bool`?

Done

> what if some of these pointers are `NULL`? I think the answer is that this isn't possible, right?

True, initialization should fail at the start (first initialization). It's definitely the case with the scheduler.

> that checking just one of these pointers should be enough: if one of them is set, all the others must be set too?

I think one would be enough, but more is better, right?

Reply:
> I think one would be enough, but more is better, right?

@tmleman why and how? *Maybe* it is better, or maybe not. If you claim, that it's better, then I'd like an explanation of why and how and a proper handling of it. So, if you detect that one of them is `NULL` and others aren't, this can mean that the initialisation code got broken. So I would expect an appropriate handling of such a case. Currently a case of some `NULL` and some non-`NULL` seems to indicate some inconsistent state and we leave it at it. A better solution could be adding an explicit "initialisation complete" flag and checking it, instead of using some implicit indicators

Reply:
> this can mean that the initialisation code got broken.

I think you have given the best answer to the question why more is better.

Perhaps it would be helpful to add a piece of code that would handle the case where only part of the components has been initialized. As for the idea with the variable (flag) that would indicate that the initialization has been completed, I don't see the value in it at the moment. What would be the difference compared to checking the state of just one component, for example, one that is initialized last?

At this moment, fixing regressions introduced by the previous refactor is a higher priority for me than starting another one.